### ISSUE TITLE
Redirect images to raw.github.com

### ISSUE THREAD

**Issue by rgrove** [rgrove/rawgit#2]
raw.github.com seems to serve the correct content types for images, so there's no need to proxy those. Should investigate other non-text filetypes as well.

**Comment by paulirish** [comment]
@rgrove I believe I just ran into this...

Over on https://github.com/siddharthkp/bundlesize/pull/126 I had suggested using cdn.rawgit rather than raw.ghusercontent but it redirects everytime. (The resource is a png).  

Is it just the case that images (and unclear mimetypes a la #87) are redirected regardless?

My usecase: For whatever reason, ghusercontent was giving a `503 Backend is unhealthy` response for a random png. So I figured going via the stackpath CDN would work out better.

![image](https://user-images.githubusercontent.com/39191/29881985-95cf06b8-8d60-11e7-87b1-9131b7322c12.png)

**Comment by rgrove** [comment]
Hey Paul!

The short answer is that, yes, both rawgit.com and cdn.rawgit.com will redirect PNGs (and other image requests) to raw.githubusercontent.com, because raw.ghuc already serves them with the correct content types.

The longer answer:

In most cases, cdn.rawgit.com will serve and cache files even when raw.ghuc already serves them with the correct content type. However, as usage of RawGit has increased, misuse and abuse have also increased, so there are some exceptions.

In the last year or so there have been several service degradations caused by people who created GitHub repos full of thousands of images or other large binary files and then directed huge amounts of traffic at cdn.rawgit.com. Normally it wouldn't be a problem for the CDN to handle massive load, but when that load is globally distributed and involves thousands of unique files it's a problem, because each CDN edge node still ends up making at least two requests to the RawGit origin server for each individual file before caching it.

In the worst cases, this led to resource exhaustion on the origin server as it tried to cope with all these requests, which degraded the service for everyone else. And for some reason these incidents tend to happen at 3am when I'm asleep and can't deal with them, which really sucks for RawGit users in Europe, Asia, and Oceania. 🙁 

Initially I would just block or throttle these users, but they would often create new GitHub accounts and do it again. So I decided to start having the CDN redirect certain file types directly to raw.ghuc, which has so far been effective in preventing problems like this.

You can see the current list of file extensions the CDN redirects in the [`extensionBlacklist` property of the config file](https://github.com/rgrove/rawgit/blob/df9a6d00f40bacb9a5a3e07d99d3785d3327c2f8/conf/index.js#L28-L48). For the most part I only add extensions to this blacklist if they've actually been a problem.

When raw.ghuc is returning errors, though, RawGit's ability to mask them is limited even without a redirect. StackPath has dozens of globally distributed edge nodes that each maintain their own cache, so even if users in Seattle were to get a cached file from cdn.rawgit.com, users in Frankfurt might get the GitHub 503. In general, if GitHub is broken, there's a good chance cdn.rawgit.com will also be broken for at least some users.

Sorry for the inconvenience, but I hope this explanation has helped!

**Comment by neoascetic** [comment]
Just my 2 cents: since I haven't faced the issues described by @rgrove yet, I do not redirect stuff on CDN URLs on [rawgithack](http://raw.githack.com).

**Comment by getify** [comment]
Question: do you consider ghusercontent to be an equally (or more!) stable/globally available CDN compared to StackPath (and RawGit)?

The reason I ask is, I am concerned about this redirect back to ghusercontent. For one, the performance of the redirect itself it an unfortunate delay... which I guess is avoidable if we just know not to use rawgit URLs for any images, but that kinda sucks.

But I'm even more concerned about relying on ghuc instead of a dedicated CDN. I dunno, maybe github's already got a great CDN on that, but I haven't found any info about it. I like and want to use RawGit because I like the peace of mind of a good CDN.

Curious your thoughts on this, and if you would ever reconsider this decision based on not being able to use a CDN with images (one of the main things you want a CDN for)?

**Comment by rgrove** [comment]
@getify Great question!

> do you consider ghusercontent to be an equally (or more!) stable/globally available CDN compared to StackPath (and RawGit)?

There's a short answer and a long answer to this question. 😄 

The short answer is that, on average, RawGit's reliability and githubusercontent.com's reliability are roughly the same, since the former depends on the latter.

The longer answer is that GitHub experiences more outages than RawGit does because it's a much more complex service, and GitHub outages tend to affect githubusercontent.com and as well. But since cdn.rawgit.com caches things indefinitely, there's a chance that RawGit could shield you from a GitHub outage, although it's not a _great_ chance, since it's entirely dependent on whether something happens to be in the cache of a particular CDN node near the requester.

> The reason I ask is, I am concerned about this redirect back to ghusercontent. For one, the performance of the redirect itself it an unfortunate delay... which I guess is avoidable if we just know not to use rawgit URLs for any images, but that kinda sucks.

I agree the redirect latency isn't ideal. But hopefully it encourages performance-conscious people not to use RawGit URLs for images.

> But I'm even more concerned about relying on ghuc instead of a dedicated CDN. I dunno, maybe github's already got a great CDN on that, but I haven't found any info about it. I like and want to use RawGit because I like the peace of mind of a good CDN.

I have no non-public knowledge of GitHub's infrastructure, but githubusercontent.com seems to be fronted by Fastly. Fastly is a fantastic CDN, so I wouldn't worry about that side of things. I'm not sure how long GitHub instructs Fastly to cache things though, and if the cache time is very short-lived, that could put requests at the mercy of whatever origin infrastructure is behind githubusercontent.com, which is a black box to me.

> Curious your thoughts on this, and if you would ever reconsider this decision based on not being able to use a CDN with images (one of the main things you want a CDN for)?

I never intended RawGit to be a general purpose free CDN, but it kind of became one because I took a wait-and-see attitude to how people were using it.

The primary purpose of RawGit always was and still is to serve HTML, JavaScript, and CSS from GitHub with correct `Content-Type` headers for quick and dirty projects. GitHub already serves images and most other binary files correctly, so RawGit is an unnecessary middleman there.

I've always felt uneasy about people relying on RawGit in production for important stuff, and I assumed people would mostly use it for demos, prototypes, and open source projects, but it turns out a _lot_ of people use it for real actual important public-facing production websites just because it's free and a real CDN would cost money.

I'm fine with taking a best-effort approach to this kind of usage as long as people are aware of the risks, but I don't plan to adapt RawGit to the needs of well-funded entities that just see it as a way to save cash by not paying for a real CDN.

Making RawGit good at serving images at scale would require more time and money, and it doesn't feel necessary to me. GitHub already serves images for free (but perhaps not with the best stability or performance). And if stability and performance are important, it's not too expensive these days to put a real, production-ready CDN like Fastly or Cloudflare or StackPath in front of files on S3.

So, tl;dr, I don't envision reconsidering this image decision anytime soon. I hope this helps explain why.

**Comment by getify** [comment]
I have another thing to point out that's quite unfortunate as a result of this situation. Compare the differences in response headers:

From **cdn.rawgit.com:**
```
access-control-allow-origin:*
cache-control:max-age=315569000, immutable
content-encoding:gzip
content-type:text/plain;charset=utf-8
date:Sat, 10 Feb 2018 00:42:22 GMT
etag:W/"f5c7faebb9f634d0e2f366ac069f6b03bcef8251"
rawgit-cache-status:MISS
server:NetDNA-cache/2.2
status:200
vary:Accept-Encoding
x-cache:MISS
x-content-type-options:nosniff
x-robots-tag:none
```

From **raw.githubusercontent.com:**
```
Accept-Ranges:bytes
Access-Control-Allow-Origin:*
Cache-Control:max-age=300
Content-Length:75880
Content-Security-Policy:default-src 'none'; style-src 'unsafe-inline'; sandbox
Content-Type:image/jpeg
Date:Sat, 10 Feb 2018 00:39:12 GMT
ETag:"c5a9e46a17b0ed90fcc45cb1c7a8536c743315ed"
Expires:Sat, 10 Feb 2018 00:44:12 GMT
Source-Age:0
Vary:Authorization,Accept-Encoding
Via:1.1 varnish
X-Cache:MISS
X-Cache-Hits:0
X-Content-Type-Options:nosniff
X-Fastly-Request-ID:6b4dc4f123678fc2844c7d1b01493431b622374d
X-Frame-Options:deny
X-Geo-Block-List:
X-GitHub-Request-Id:FEEE:2CB3:1F9942:2142B0:5A7E3F30
X-Served-By:cache-mdw17324-MDW
X-Timer:S1518223152.343583,VS0,VE68
X-XSS-Protection:1; mode=block
```

Specifically, look at the differences in caching. Both URLs were using a commit URL (so basically immutable). **cdn.rawgit.com** served up an "immutable" cache-control with a 1 year max-age. **raw.githubusercontent.com** served up cache-control and expirations of 5 minutes. Ouch.

So, yeah... using **raw.githubusercontent.com** as a CDN is not a good idea. And since you're not willing to serve those, I'm kinda out of luck. :( Well, I guess precisely, the users of my site are going to have quite suboptimal performance.

**Comment by rgrove** [comment]
@getify Have you considered using a free Cloudflare or Netlify plan? If you're hosting content for an open source project, I know Fastly also offers free CDN services for open source projects. There are many free and low-cost CDN options these days.

**Comment by getify** [comment]
I'm not aware of how to set any of that up. I'm not hosting the files in question, they're on a github repo. So I don't know how to set up some CDN to point at (cache from) files on github. Ostensibly, that's what **raw.githubusercontent.com** and/or **cdn.rawgit.com** seemed explicitly to be for, and best at.

**Comment by rgrove** [comment]
[This guide](https://blog.cloudflare.com/secure-and-fast-github-pages-with-cloudflare/) might be helpful as a starting place if using GitHub Pages is an option.

**Comment by getify** [comment]
I'm not positive, but I don't think GHPages is an option because the design of my service relies on the permanent commit hash URLs to the images. IIUC, GHPages always points to HEAD.

----

Anyway, I understand why you're not wanting to serve images (because you got burned). And I also understand that you set about only to create a service to serve correct mimetypes, and not really a general CDN. But... you did call it a CDN, and you did engage a real CDN (StackPath) to do the endpoint serving. That's where my broken expectations came from. :(

I wish you would reconsider, so I don't have to maybe contact StackPath myself and try to reinvent what you've already done. But I guess I understand your stance.

**Comment by rgrove** [comment]
Worth noting: RawGit is an open source Node.js app. If you have a [$5/month DigitalOcean droplet](https://www.digitalocean.com/pricing/) and a free Cloudflare account, you can run your own RawGit instance and have it serve anything you need, fronted by a top-notch CDN. Maybe you can even skip the $5 a month and run RawGit on a free [Now deployment](https://zeit.co/now).

You may have to spend some time setting it up and learning how things work, but if what you're doing isn't worth your time, then my advice is don't do it. 😄 

> But... you did call it a CDN, and you did engage a real CDN (StackPath) to do the endpoint serving. That's where my broken expectations came from. :(

I resisted adding an actual CDN for a long time. StackPath (then MaxCDN) approached me twice. I didn't approach them.

The first time I turned them down. The second time I accepted their generous offer to provide free CDN service because users were already using RawGit as if it was a CDN and it was becoming more and more difficult to keep the site up. If I hadn't added the CDN, I think I would've had to shut RawGit down a year or two after that point simply due to an inability to handle the traffic people were throwing at it.

The reality is that these days I have, at most, an hour or two a month that I can devote to RawGit. I also lead the engineering effort for an ambitious startup, which consumes an enormous amount of my stress budget. I run RawGit with my own money and with StackPath's generously donated CDN service. And RawGit's traffic is still growing, so it's not likely to get any easier.

I sympathize with people who want RawGit to do more, but I want to set realistic expectations: RawGit will not do more. If anything, it will do less. If it doesn't do what you want it to do, you will need to invest some of your own time and/or money in accomplishing those things.

**Comment by neoascetic** [comment]
@getify take a look at rawcdn.githack.com, as I [said earlier](https://github.com/rgrove/rawgit/issues/2#issuecomment-326077434), I do not redirect any stuff on CDN URLs (as it going to be cached for a long time anyway), for example: 

![](https://raw.githubusercontent.com/rgrove/rawgit/cdn-20171119/public/img/sushi.png)

CDN URLs are backed by CloudFlare CDN, so you could try it out.
